posthog-core/src/index.ts

```diff
@@ -63,38 +57,422 @@ export abstract class PostHogCore {
+    this.setPersistedProperty<PostHogQueueItem[]>(PostHogPersistedProperty.Queue, queue)
+
+    const messages = items.map((item) => item.message)
+
+    const data = {
+      api_key: this.apiKey,
+      batch: messages,
+      sent_at: currentISOTime(),
+    }
+
+    const done = (err?: any): void => {
+      callback?.(err, messages)
+      this._events.emit('flush', messages)
+    }
+
+    // Don't set the user agent if we're not on a browser. The latest spec allows
+    // the User-Agent header (see https://fetch.spec.whatwg.org/#terminology-headers
+    // and https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/setRequestHeader),
+    // but browsers such as Chrome and Safari have not caught up.
+    const customUserAgent = this.getCustomUserAgent()
+    const headers: { [key: string]: string } = {}
+    if (customUserAgent) {
+      headers['user-agent'] = customUserAgent
+    }
+
+    const payload = JSON.stringify(data)
+
+    const url =
+      this.captureMode === 'form'
+        ? `${this.host}/e/?ip=1&_=${currentTimestamp()}&v=${this.getLibraryVersion()}`
+        : `${this.host}/batch/`
+
+    const fetchOptions: PostHogFetchOptions =
+      this.captureMode === 'form'
+        ? {
+            method: 'POST',
+            mode: 'no-cors',
+            credentials: 'omit',
+            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
+            body: `data=${encodeURIComponent(LZString.compressToBase64(payload))}&compression=lz64`,
+          }
+        : {
+            method: 'POST',
+            headers: { 'Content-Type': 'application/json' },
+            body: payload,
+          }
+
+    this.fetchWithRetry(url, fetchOptions)
+      .then(() => done())
+      .catch((err) => {
+        if (err.response) {
+          const error = new Error(err.response.statusText)
+          return done(error)
+        }
+
+        done(err)
+      })
+  }
+
+  private async fetchWithRetry(
+    url: string,
+    options: PostHogFetchOptions,
+    retryOptions?: RetriableOptions
+  ): Promise<PostHogFetchResponse> {
+    ;(AbortSignal as any).timeout ??= function timeout(ms: number) {
+      const ctrl = new AbortController()
+      setTimeout(() => ctrl.abort(), ms)
+      return ctrl.signal
+    }
+
+    return retriable(
+      () =>
+        this.fetch(url, {
+          signal: (AbortSignal as any).timeout(this.requestTimeout),
+          ...options,
+        }),
+      retryOptions || this._retryOptions
+    )
+  }
+
+  async shutdownAsync(): Promise<void> {
+    clearTimeout(this._flushTimer)
+    await this.flushAsync()
+  }
+
+  shutdown(): void {
+    void this.shutdownAsync()
+  }
+}
+
+export abstract class PostHogCore extends PostHogCoreStateless {
+  // options
+  private sendFeatureFlagEvent: boolean
+  private flagCallReported: { [key: string]: boolean } = {}
+
+  // internal
+  protected _decideResponsePromise?: Promise<PostHogDecideResponse | undefined> // TODO: come back to this, fix typing
```

Comment:
Still todo?

Reply:
yes-ish, want to confirm request structure in a proper example with node once it's in place. Can I leave this in for the follow up? 😅